hail/src/test/scala/is/hail/io/compress/BGzipCodecSuite.scala

```diff
@@ -114,4 +115,90 @@ class BGzipCodecSuite extends SparkSuite {
     }
     p.check()
   }
+
+  @Test def testVirtualSeek() {
+    // real offsets of the start of some blocks, paired with the offset to the next block
+    val blockStarts = Array[(Long, Long)]((0, 14653), (69140, 82949), (133703, 146664), (181362, 192983 /* end of file */))
+    val maxBlockSize = 65280
+    // number determined by counting bytes from sample.vcf from uncompBlockStarts.last to the end of the file
+    val lastBlockLen = 55936
+    // offsets into the uncompressed file
+    val uncompBlockStarts = Array[Int](0, 326400, 652800, 913920)
+
+    val uncompPath = "src/test/resources/sample.vcf"
+    val compPath = "src/test/resources/sample.vcf.gz"
+
+    val uncompHPath = new hd.fs.Path(uncompPath)
+    val compHPath = new hd.fs.Path(compPath)
+
+    val fs = uncompHPath.getFileSystem(hadoopConf)
+    val compIS = fs.open(compHPath)
+
+    using (fs.open(uncompHPath)) { uncompIS => using (new BGzipInputStream(compIS)) { decompIS =>
+      val fromEnd = 48 // arbitrary number of bytes from the end of block to attempt to seek to
+      for (((cOff, nOff), uOff) <- blockStarts.zip(uncompBlockStarts);
+           e <- Seq(0, 1024, maxBlockSize - fromEnd)) {
+        val decompData = new Array[Byte](100)
+        val uncompData = new Array[Byte](100)
+        val extra = if (cOff == blockStarts.last._1 && e == maxBlockSize - fromEnd)
+            lastBlockLen - fromEnd
+          else
+            e
+        val vOff = BlockCompressedFilePointerUtil.makeFilePointer(cOff, extra)
+
+        decompIS.virtualSeek(vOff)
+        assert(decompIS.getVirtualOffset() == vOff);
+        uncompIS.seek(uOff + extra)
+
+        val decompRead = {
+          var r = decompIS.read(decompData)
+          var rd = 1
+          while (r < decompData.length && rd > 0) {
+            rd = decompIS.read(decompData, r, decompData.length - r)
+            if (rd > 0)
+              r += rd
+          }
+          r
+        }
+        val uncompRead = uncompIS.read(uncompData)
+
+        assert(decompRead == uncompRead, s"""compressed offset: ${ cOff }
+          |decomp bytes read: ${ decompRead }
+          |uncomp bytes read: ${ uncompRead }\n""".stripMargin)
+        assert(decompData sameElements uncompData, s"data differs for compressed offset: ${ cOff }")
+        val expectedVirtualOffset = if (extra == lastBlockLen - fromEnd)
+            BlockCompressedFilePointerUtil.makeFilePointer(nOff, 0)
+          else if (extra == maxBlockSize - fromEnd)
+            BlockCompressedFilePointerUtil.makeFilePointer(nOff, decompRead - fromEnd)
+          else
+            BlockCompressedFilePointerUtil.makeFilePointer(cOff, extra + decompRead)
+        assert(expectedVirtualOffset == decompIS.getVirtualOffset())
+      }
+
+      // Trying to seek to the end of a block should fail
+      intercept[java.io.IOException] {
+        val vOff = BlockCompressedFilePointerUtil.makeFilePointer(blockStarts(1)._1, maxBlockSize);
```

Comment:
Is maxBlockSize the size of the second block?  That's not obvious.

Reply:
Yes, I'll add a comment